docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+## Use cases
+
+We've identified several first and third party candidates that would benefit from source generators:
+
+- ASP.Net: Improve startup time 
+- Blazor and Razor: Massively reduce tooling burden 
+- Azure Functions: regex compilation during startup
+- [gRPC](https://docs.microsoft.com/en-us/aspnet/core/grpc/?view=aspnetcore-3.1)
+- Resx file generation
+- [System.CommandLine](https://github.com/dotnet/command-line-api)
+- Serializers
+- [SWIG](http://www.swig.org/)
+
+
+## Discussion / Open Issues / TODOs:
+
+**Interface vs Class for ISourceGenerator**: 
+
+We discussed about this being an interface or class. Analyzers chose to have a abstract base class,
+but we weren't sure what we'd end up a need since ultimately we only had one method on this.
+Keeping it an interface also was more natural since we have other interfaces that 
+implement this interface as well for optional light-up.
+
+**IDependsOnCompilationGenerator**:
+
+We did discuss if there should be an IDependsOnCompilationGenerator to formally state 
+that you actually use a compilation. After all, if you don't use the compilation 
+then we know your performance in the IDE is greatly simplified. However every 
+scenario we've had for reading additional files has also needed the compilation, 
+so we simply weren't sure what that was going to bring.
+
+**Breakpoints in generated files**: 
+
+Do we map this back to the in-memory file?
+
+**Should generators be push or pull**:
+
+Source generators are pull-based, analyzers are push-based (registration based). Should
+we use a push-based model for generators as well?
+
+- If we go down the push-based model, walking the tree should make sure to continue
+to produce events for as many nodes as possible, even with errors, as generators
+will often work in the presence 
+
+- The events that we use today for analyzers may require may more work to produce,
+since we expect analyzers to run during full compilation, while generators may
+not want to even construct the symbol table
+
+- The progressive-performance-opt-in model may work better in a push-based model,
+since you would only register for the things you care about
+
+**Should we share more with the analyzer type hierarchy?**:
+
+We would still need to differentiate analyzers from generators, since
+they would be generated at different times (generator diagnostics only on
+the first compilation, analyzer diagnostics only on the second compilation)
+
+**Can we predict how often some of our sample customers (Razor?) will have to run the generators?**:
+
+They can't predict that right now, and the incorporation of timers into their
+current generation makes it very difficult to predict the consequences of
+only event-based generation
+
+**Do we have a priority list of the most important customers?**:
+
+No, we should work out priority in order to prioritize features.
+
+**Security Review**:
+
+Do generators create any new security risks not already posed via analyzers and nuget?
```

Comment:
Certainly they do introduce new security risks.  See my "insert a virus into my program at compile-time" generator.

Reply:
How is this a new security risk though? Or more accurately how is the security risk of a source generator any different than that of analyzers? A misbehaving analyzer could use reflection tricks to do virtually anything a source generator could do. Certainly a bad actor wouldn't baulk at using reflection just because it wasn't supported.

Reply:
As far as we understand, there is nothing a generator can do that an analyzer couldn't have achieved previously via private reflection. Ultimately we've been allowing arbitrary code execution from packages during compilation for some time (especially via nuget), we're just making it explicit now. This is a note that we need to chat with the security folks to ensure our assumptions are correct.

---
In reply to: [354613044](https://github.com/dotnet/roslyn/pull/40162#discussion_r354613044) [](ancestors = 354613044)

Reply:
Analyzers come via package references and package references can do basically anything. They can install props, targets, random non-.NET binaries, play the he-man song on repeat, etc ... Even in the case we assume they're checked (they're not) analyzers are code running with full trust on the machine.

Reply:
For context, this is the he-man song: http://aka.ms/davkean